### PR TITLE
ARROW-17408: [c++] Fixing C++20 compile warning using operator==(const T&, const T&)

### PR DESCRIPTION
**Pull request by mapleFU**
Report and fixing this issue via: https://stackoverflow.com/questions/65833022/inherited-synthesized-comparison-operator-produces-warning-error-iso-c20-cons

### CODE REVIEW

**Comment by kou** [comment]
It seems that this breaks VC++ CI: https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/44467048/job/2tamgyjibu9kuqvi

```text
C:\projects\arrow\cpp\src\arrow/util/compare.h(55): error C2593: 'operator ==' is ambiguous
C:\projects\arrow\cpp\src\arrow/util/compare.h(54): note: could be 'bool arrow::util::EqualityComparable<arrow::Scalar>::operator ==(const T &) const'
        with
        [
            T=arrow::Scalar
        ]
C:\projects\arrow\cpp\src\arrow/util/compare.h(57): note: or       'bool arrow::util::operator ==(const T &,const T &)' [found using argument-dependent lookup]
        with
        [
            T=arrow::Scalar
        ]
C:\projects\arrow\cpp\src\arrow/util/compare.h(55): note: while trying to match the argument list '(const T, const T)'
        with
        [
            T=arrow::Scalar
        ]
C:\projects\arrow\cpp\src\arrow/util/compare.h(55): note: while compiling class template member function 'bool arrow::util::EqualityComparable<arrow::Scalar>::operator !=(const T &) const'
        with
        [
            T=arrow::Scalar
        ]
C:\projects\arrow\cpp\src\arrow/scalar.h(53): note: see reference to class template instantiation 'arrow::util::EqualityComparable<arrow::Scalar>' being compiled
```

Could you also add a C++20 CI job?

```diff
diff --git a/dev/tasks/tasks.yml b/dev/tasks/tasks.yml
index 7ecc610224..068c956dc8 100644
--- a/dev/tasks/tasks.yml
+++ b/dev/tasks/tasks.yml
@@ -1211,7 +1211,7 @@ tasks:
       flags: "-e ARROW_BUILD_SHARED=OFF -e ARROW_BUILD_STATIC=ON -e ARROW_TEST_LINKAGE=static"
       image: ubuntu-cpp
 
-{% for cpp_standard in [14, 17] %}
+{% for cpp_standard in [14, 17, 20] %}
   test-ubuntu-20.04-cpp-{{ cpp_standard }}:
     ci: github
     template: docker-tests/github.linux.yml
```

**Comment by mapleFU** [comment]
@kou Yes, I will fix it later

File: cpp/src/arrow/util/compare.h [hunk 1]
```diff
@@ -54,6 +54,10 @@ class EqualityComparable {
   bool operator==(const T& other) const { return cast().Equals(other); }
   bool operator!=(const T& other) const { return !(cast() == other); }
 
+  friend bool operator==(const T& a, const T& b) {
+    return a.Equals(b);
+  }
```
Comment (frederick-vs-ja) [inline cpp/src/arrow/util/compare.h:59]: Perhaps we can only use hidden friends.

```suggestion
  friend bool operator==(const T& a, const T& b) {
    return a.Equals(b);
  }
  friend bool operator!=(const T& a, const T& b) {
    return !(a.Equals(b));
  }
```

Comment (mapleFU) [reply cpp/src/arrow/util/compare.h:59]: It's ok on my PC, I'll try it

**Comment by mapleFU** [comment]
@kou Hi, my fixing works for C+20, but would not compile in some platforms. Should I add some macros that make it only works in C+20? Or can you give me some advice? Thanks a lot

**Comment by kou** [comment]
@pitrou Do you have any idea to support C++20 and earlier?

**Comment by pitrou** [comment]
@kou Can we wait for the C++17 switch proposal to be accepted and revisit this?

**Comment by kou** [comment]
Sure.

Reference: DISCUSS: [C++] Switch to C++17
https://lists.apache.org/thread/9g14n3odhj6kzsgjxr6k6d3q73hg2njr

**Comment by pitrou** [comment]
This could be rebased from master now that we switched to C++17.

**Comment by pitrou** [comment]
@mapleFU Are you interested in updating this PR?

**Comment by mapleFU** [comment]
> Are you interested in updating this PR?

So busy these days, will continue on it tonight

**Comment by frederick-vs-ja** [comment]
I think the original issue has been resolved in the Standard: [P2468R2](https://wg21.link/p2468r2) has been accepted as a Defect Report against C++20, so compilers should consider the use of `arrow::util::EqualityComparable` still valid in C++20 mode.

This DR has been implemented in gcc 13 and clang 16. Although old versions (perhaps together with the current msvc) still emit warnings or errors.

**Pull request closed without merging**